pkg/volume/downwardapi/downwardapi.go

```diff
@@ -0,0 +1,329 @@
+// 4.  In case no symlink exists then it's created and func returns (first execution)
+// 5.  In case symlink exists a new temporary symlink is created .current_tmp
+// 6.  .current_tmp is renamed to .current
+// 7.  The previous timestamped directory is removed
+func (m *downwardAPIVolume) writeData(data *map[string]string) error {
+	var timestampDir string
+	var err error
+	timestampDir, err = m.writeDataInTimestampDir(data)
+	if err != nil {
+		glog.Error(err)
+		return err
+	}
+	// update symbolic links for relative paths
+	if err = m.updateSymlinksToCurrentDir(); err != nil {
+		os.RemoveAll(timestampDir)
+		glog.Error(err)
+		return err
+	}
+
+	_, timestampDirBaseName := filepath.Split(timestampDir)
+	var oldTimestampDirectory string
+	oldTimestampDirectory, err = os.Readlink(path.Join(m.GetPath(), currentDir))
+	if err != nil {
+		if os.IsNotExist(err) { // no link to currentDir so creates it
+			err = os.Symlink(timestampDirBaseName, path.Join(m.GetPath(), currentDir))
+		}
+		return err // in any cases return
+	}
+
+	// nominal case: link to currentDir exists create the link to currentTmpDir
+	if err = os.Symlink(timestampDirBaseName, path.Join(m.GetPath(), currentTmpDir)); err != nil {
+		os.RemoveAll(timestampDir)
+		glog.Error(err)
+		return err
+	}
+
+	// Rename the symbolic link currentTmpDir to currentDir
+	if err = os.Rename(path.Join(m.GetPath(), currentTmpDir), path.Join(m.GetPath(), currentDir)); err != nil {
+		// in case of error remove latest data and currentTmpDir
+		os.Remove(path.Join(m.GetPath(), currentTmpDir))
+		os.RemoveAll(timestampDir)
+		glog.Error(err)
+		return err
+	}
+	// Remove oldTimestampDirectory
+	if len(oldTimestampDirectory) > 0 {
+		if e := os.RemoveAll(path.Join(m.GetPath(), oldTimestampDirectory)); e != nil {
+			glog.Error(e)
+			return e
+		}
+	}
+	return nil
+}
+
+// writeDataInTimestampDir writes the latest data into a new temporary directory with a timestamp.
+func (m *downwardAPIVolume) writeDataInTimestampDir(data *map[string]string) (string, error) {
+	errlist := []error{}
+	timestampDir, err := ioutil.TempDir(m.GetPath(), "."+time.Now().Format("2006_01_02_15_04_05"))
+	for fileName, values := range *data {
+		fullPathFile := path.Join(timestampDir, fileName)
+		dir, _ := filepath.Split(fullPathFile)
+		if err = os.MkdirAll(dir, os.ModePerm); err != nil {
+			return "", err
+		}
+		if e := ioutil.WriteFile(fullPathFile, []byte(values), 0644); e != nil {
+			glog.Error(e)
+			errlist = append(errlist, err)
+		}
+	}
+	return timestampDir, utilErrors.NewAggregate(errlist)
+}
+
+// updateSymlinksToCurrentDir creates the relative symlinks for all the files configured in this volume.
+// If the directory in a file path does not exist, it is created.
+func (m *downwardAPIVolume) updateSymlinksToCurrentDir() error {
+	for _, f := range m.fieldReferenceFileNames {
+		dir, _ := filepath.Split(f)
+		nbOfSubdir := 0
+		if len(dir) > 0 {
+			nbOfSubdir = len(strings.Split(dir, "/")) - 1
```

Comment:
This will break if anyone says "foo//bar" - you should call CleanPath() on the path names when you originally save them, so that this becomes safe.